examples/adversarial_training.ipynb

```diff
@@ -0,0 +1,1052 @@
+    {
+      "cell_type": "code",
+      "execution_count": 6,
+      "metadata": {
+        "id": "trBo7PpcQ4N8",
+        "outputId": "aa85ee43-aabc-4545-d753-a670bd3dbbf5",
+        "colab": {
+          "base_uri": "https://localhost:8080/",
+          "height": 84,
+          "referenced_widgets": [
+            "e1c2e68ebfde40d0a378d864591454ae",
+            "0809e931a2cf4d79a0b4f4adc74cc0a6",
+            "446a3d8753bc4518afc6931dc7a04036",
+            "0826f6b966d0462b93340f98cd0def17",
+            "0bae2a46b67742c681f9d5d2e438be08",
+            "27ce1b1ca61e44df8de592a60713aa19",
+            "d8eec547f16943c99f3c501fca38a927",
+            "dfb4d7931ae14f11bc47d0d8508daee5",
+            "5d7a7618e370456480654b4f1dfbdda2",
+            "cede9d5af636471ab98ffe3e48a275df",
+            "0d2e55eabc6041479bf88a6fbecff7f4"
+          ]
+        }
+      },
+      "outputs": [
+        {
+          "output_type": "stream",
+          "name": "stdout",
+          "text": [
+            "Downloading and preparing dataset 11.06 MiB (download: 11.06 MiB, generated: 21.00 MiB, total: 32.06 MiB) to /root/tensorflow_datasets/mnist/3.0.1...\n"
+          ]
+        },
+        {
+          "output_type": "display_data",
+          "data": {
+            "text/plain": [
+              "Dl Completed...:   0%|          | 0/5 [00:00<?, ? file/s]"
+            ],
+            "application/vnd.jupyter.widget-view+json": {
+              "version_major": 2,
+              "version_minor": 0,
+              "model_id": "e1c2e68ebfde40d0a378d864591454ae"
+            }
+          },
+          "metadata": {}
+        },
+        {
+          "output_type": "stream",
+          "name": "stdout",
+          "text": [
+            "Dataset mnist downloaded and prepared to /root/tensorflow_datasets/mnist/3.0.1. Subsequent calls will reuse this data.\n"
+          ]
+        }
+      ],
+      "source": [
+        "train_ds, ds_info = load_dataset(\"train\", is_training=True,\n",
+        "                                 batch_size=FLAGS.train_batch_size)\n",
+        "test_ds, _ = load_dataset(\"test\", is_training=False,\n",
+        "                          batch_size=FLAGS.test_batch_size)\n",
+        "input_shape = (1,) + ds_info.features[\"image\"].shape\n",
+        "num_classes = ds_info.features[\"label\"].num_classes\n",
+        "iter_per_epoch_train = ds_info.splits['train'].num_examples // FLAGS.train_batch_size\n",
+        "iter_per_epoch_test = ds_info.splits['test'].num_examples // FLAGS.test_batch_size\n",
+        "\n",
+        "\n",
+        "net = CNN(num_classes)\n",
+        "\n",
+        "@jax.jit\n",
+        "def accuracy(params, data):\n",
+        "  inputs, labels = data\n",
+        "  logits = net.apply({\"params\": params}, inputs)\n",
+        "  return jnp.mean(jnp.argmax(logits, axis=-1) == labels)\n",
+        "\n",
+        "@jax.vmap\n",
+        "def logistic_loss(label: int, logits: jnp.ndarray) -> float:\n",
```

Comment:
please use one of optax's predefined losses (in this case I think it's https://optax.readthedocs.io/en/latest/api.html#optax.softmax_cross_entropy_with_integer_labels )